### DISCUSSION TITLE
Does the JanusGraph database support predicate pushdown?

### DISCUSSION THREAD

**Discussion by yoylee** [JanusGraph/janusgraph#2286] [Q&A] [answered]
For example, let's imagine we have a vertex label user and edge label follows. Now let's imagine we have a vertex user (user A) who is following 10,000 other users and our business case is to return time is December 18 users . Each edge follows contains a property time (let's assume it's unix timestamp).

A query:`g.V(id).outE("follows").has('time', '1608258863')`

Is this part <`has('time', '1608258863')`> filtered in the underlying hbase or transfer to janusgraph to filter?

thanks

**Comment by li-boxuan** [comment, marked as answer]
If you have a vertex-centric index for 'time' property then no filtering is needed, otherwise in-memory filtering is done in JanusGraph.

**Reply by simon824** [reply]
it means if there is  a vertex-centric index , it will be filtered in the underlying hbase by ColumnFilter？

**Reply by li-boxuan** [reply]
Yes

**Reply by porunov** [reply]
@li-boxuan answered correctly.
edge filer step i.e. (the above use case .has('time', 12345)) will be filtered using index on the underlying storage backend, so no in memory filtering will be conducted.
If you don't have vertex centric index covering your use case, JanusGraph will be required to fetch edges in memory and conduct filtering.
Vertex centric indexes also support predicates. For example, if you want to return edges with time between 12345 and 54321 you can write your query like: 
`g.V(id).outE("follows").has('time', P.between(12345, 54321))`. The previous query will use vertex centric index if you have OUT or BOTH vertex centric index on `time`.
